File: 2022/day25/zig/solution.zig
const std = @import("std");

fn snafuToDecimal(s: []const u8) i64 {
    var result: i64 = 0;
    for (s) |c| {
        const digit_value: i64 = switch (c) {
            '2' => 2,
            '1' => 1,
            '0' => 0,
            '-' => -1,
            '=' => -2,
            else => 0,
        };
        result = result * 5 + digit_value;
    }
    return result;
}

fn decimalToSnafu(n: i64, buffer: []u8) []const u8 {
    if (n == 0) {
        buffer[0] = '0';
        return buffer[0..1];
    }

    var digits: [32]u8 = undefined;
    var digit_count: usize = 0;
    var num = n;

    while (num != 0) {
        const remainder: i64 = @mod(num, 5);
        if (remainder <= 2) {
            digits[digit_count] = @intCast('0' + @as(u8, @intCast(remainder)));
            num = @divFloor(num, 5);
        } else if (remainder == 3) {
            digits[digit_count] = '=';
            num = @divFloor(num, 5) + 1;
        } else { // remainder == 4
            digits[digit_count] = '-';
            num = @divFloor(num, 5) + 1;
        }
        digit_count += 1;
    }

    // Reverse the digits into the output buffer
    for (0..digit_count) |i| {
        buffer[i] = digits[digit_count - 1 - i];
    }

    return buffer[0..digit_count];
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    // Read input file
    const file = std.fs.cwd().openFile("../input.txt", .{}) catch |err| {
        std.debug.print("Failed to open ../input.txt: {}\n", .{err});
        return err;
    };
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Parse and sum all SNAFU numbers
    var total: i64 = 0;
    var lines = std.mem.splitScalar(u8, content, '\n');
    while (lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, &[_]u8{ '\r', ' ' });
        if (trimmed.len > 0) {
            total += snafuToDecimal(trimmed);
        }
    }

    // Convert back to SNAFU
    var snafu_buffer: [32]u8 = undefined;
    const result = decimalToSnafu(total, &snafu_buffer);

    // Use posix write to stdout
    const stdout_fd = std.posix.STDOUT_FILENO;
    var buf1: [64]u8 = undefined;
    const msg1 = std.fmt.bufPrint(&buf1, "Part 1: {s}\n", .{result}) catch unreachable;
    _ = std.posix.write(stdout_fd, msg1) catch {};

    var buf2: [64]u8 = undefined;
    const msg2 = std.fmt.bufPrint(&buf2, "Part 2: No Part 2 on Day 25!\n", .{}) catch unreachable;
    _ = std.posix.write(stdout_fd, msg2) catch {};
}

File: 2022/day25/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 64
#define MAX_LINES 256

/* Convert a SNAFU digit to its decimal value */
int snafu_digit_value(char c) {
    switch (c) {
        case '2': return 2;
        case '1': return 1;
        case '0': return 0;
        case '-': return -1;
        case '=': return -2;
        default: return 0;
    }
}

/* Convert SNAFU string to decimal */
long long snafu_to_decimal(const char *s) {
    long long result = 0;
    for (int i = 0; s[i] != '\0' && s[i] != '\n'; i++) {
        result = result * 5 + snafu_digit_value(s[i]);
    }
    return result;
}

/* Convert decimal to SNAFU string (stores result in buffer) */
void decimal_to_snafu(long long n, char *result) {
    if (n == 0) {
        strcpy(result, "0");
        return;
    }

    char digits[MAX_LINE];
    int idx = 0;

    while (n > 0) {
        int remainder = n % 5;
        if (remainder <= 2) {
            digits[idx++] = '0' + remainder;
            n /= 5;
        } else if (remainder == 3) {
            digits[idx++] = '=';
            n = n / 5 + 1;
        } else { /* remainder == 4 */
            digits[idx++] = '-';
            n = n / 5 + 1;
        }
    }

    /* Reverse the digits into result */
    for (int i = 0; i < idx; i++) {
        result[i] = digits[idx - 1 - i];
    }
    result[idx] = '\0';
}

int main(void) {
    FILE *fp = fopen("../input.txt", "r");
    if (!fp) {
        perror("Failed to open input.txt");
        return 1;
    }

    char line[MAX_LINE];
    long long total = 0;

    while (fgets(line, sizeof(line), fp)) {
        if (line[0] == '\n' || line[0] == '\0') continue;
        total += snafu_to_decimal(line);
    }

    fclose(fp);

    char snafu_result[MAX_LINE];
    decimal_to_snafu(total, snafu_result);

    printf("Part 1: %s\n", snafu_result);
    printf("Part 2: No Part 2 on Day 25!\n");

    return 0;
}
